Blog post: Migrate Ember app to CSS modules (#1838)
simplabs/simplabs.github.io: _posts/2022-08-05-cookbook-ember-app-to-css-modules.md

---
title: 'Cookbook: migrate an existing Ember app to CSS modules'
author: 'Marine Dunstetter'
github: BlueCutOfficial
twitter: academierenards
topic: open-source
bio: ''
description: ''
og:
  image: /assets/images/posts/2022-08-05-cookbook-ember-app-to-css-modules/og-image.png
---

This short tutorial is intended to help Ember developers to migrate an Ember app
using only one global CSS to CSS modules with
[`ember-css-modules`](https://github.com/salsify/ember-css-modules).

<!--break-->

## Introduction & prerequisites

[`ember-css-modules`](https://github.com/salsify/ember-css-modules) is a nice

> addon that let developers easily integrate CSS modules to the implementation of

Author: Maybe the solution is to locate this after the prerequisite list and these references to docs about CSS modules I need to add (see general comment).

> addon that let developers easily integrate CSS modules to the implementation of

Reviewer: see my comment – one paragraph introducing the concept would be helpful (even if most readers might know already)

their Ember apps and addon. Installig `ember-css-modules` in an existing Ember
app triggers some breaking changes in the way the integrator should handle the
app style. And if the integrator is not familiar with CSS modules in the first
place, they might spend a hard time figuring out how to set up and start the
migration at their own pace. The present tutorial is intended to be used as a
quick start cookbook to achieve this.

In the next section, we suppose the following:

- You are already comfortable with Ember.
- It is the first time you set up
  [`ember-css-modules`](https://github.com/salsify/ember-css-modules) in an
  existing app that already contains a global CSS.
- You know a bit the theory about what CSS modules are but it might be the first
  time you actually try to work with them.

The usage of [less](https://lesscss.org/) and
[sass](https://sass-lang.com/guide) will not be mentioned, though they are
[compatible with `ember-css-modules`](https://github.com/salsify/ember-css-modules/blob/master/docs/PREPROCESSORS.md).

## Migration tutorial

We'll take
[SuperRentals app](https://guides.emberjs.com/release/tutorial/part-1/) as an

> example. SuperRentals is the great tutorial part of the Ember guides to

Author: I meant "the" in the way I am talking about a very specific tutorial (the SuperRentals one), but there's certainly a way to turn the whole sentence better... -_-

<div class="right tomster"></div>
```

```css
/* app/styles/app.css */

.right {
  float: right;
}

.tomster {
  background: url(../assets/images/teaching-tomster.png);
  background-size: contain;
  background-repeat: no-repeat;
  height: 200px;
  width: 200px;

  position: relative;
  top: -25px;
}
```

Now, let's have a look to the CSS applied on the `<div>` element:

![Tomster div highlighted in the DOM](/assets/images/posts/2022-08-05-cookbook-ember-app-to-css-modules/screen-3.png)

Only two CSS rules apply on the Tomster `<div>`. The first one is a global rule
to reset margins. The second one defines the font family and line height set on
elements as `body`, `h` titles, `p`, `div`, etc... The CSS classes responsible
for showing Tomster are `.right` and `.tomster`, but these classes are no longer
there, that's why the tomster image is no longer visible on screen.

In the section of the inspector dedicated to CSS description, when a style
applies on the selected element, we can see the name of the CSS file this style
comes from. On the screenshot above, it comes from `super-rental-typescript`
(this is because the final "s" on rentals was forgotten ^^' and this demo app

> was implemented as a migration test to Typescript). The file name is clickable

Reviewer 2: Should be capitalized as `TypeScript`. But is this relevant to the article?

and let you see the content of the whole CSS file.

At the top of the file, we read:

```css
* {
  margin: 0;
  padding: 0;
}

body,
h1,
h2,
h3,
h4,
h5,
h6,
p,
div,
span,
a,
button {
  font-family: 'Lato', 'Open Sans', 'Helvetica Neue', 'Segoe UI', Helvetica,
    Arial, sans-serif;
  line-height: 1.5;
}
```

At the bottom of the file, we read:

```css
._right_1s31t7 {
  float: right;
}

._tomster_1s31t7 {
  background: url(../assets/images/teaching-tomster.png);
  background-size: contain;
  background-repeat: no-repeat;
  height: 200px;
  width: 200px;

  position: relative;
  top: -25px;
}
```

And here is the explanation about our missing Tomster. With `ember-css-modules`
installed, all the class selectors present in `app.css` end with a custom id.
But this custom id is not present in our component's template, we still have
`class="right tomster"`. As result, no CSS style is found for `.right` and
`.tomster` and no style applies. At the contrary, the rules defined for main
elements like `div` don't have any custom id added and keep applying.

`ember-css-modules` relies on class selectors to isolate the CSS and prevent it

> to apply globally. The isolation system applies on every CSS files, including

Reviewer 2: prevent `from`

those located in the main `styles` folder rather than alongside a component.

### Fix the set up with `:global`

Before starting to activly use CSS modules accross the app, we would prefer to
fix the style first, and alllow `app.css` to apply globally like it did before.
To do so, we can use the `:global` directive:

```css
/* app/styles/app.css */

:global(.right) {
  float: right;
}

:global(.tomster) {
  background: url(../assets/images/teaching-tomster.png);
  background-size: contain;
  background-repeat: no-repeat;
  height: 200px;
  width: 200px;

  position: relative;
  top: -25px;
}
```

And let's see what happens in the browser:

![Tomster shows again](/assets/images/posts/2022-08-05-cookbook-ember-app-to-css-modules/screen-4.png)

Tomster is back! If we take a new look at the generated CSS file in the
browser's inspector, we can see this time the classes `.right` and `.tomster`

> don't have a custom id. The `:global` directive indicates these classes are

Reviewer 2: Technically, `:global` is not a directive and more like a pseudoselector... But it may be OK to keep using the word `directive`.

Author: I picked the word directive in the documentation, I'll read again to be sure it was the same context.

resulted to `<div class="right _tomster_1myv6d"></div>` in the browser to match
the class selector.

### One last note

As we experienced above, `class` and `local-class` can be used together to apply
global and local styles, everything in `local-class` ending with a different
selector. If it makes sense, in order to ease the migration of your app to CSS
modules, note that the following works:

```html
<div class="right tomster" local-class="tomster"></div>
```

```css
/* app/styles/app.css */

:global(.right) {
  float: right;
}

:global(.tomster) {
  width: 60px;
  height: 60px; /* micro tomster */
}
```

```css
/* app/components/jumbo.css */

.tomster {
  background: url(../assets/images/teaching-tomster.png);
  background-size: contain;
  background-repeat: no-repeat;

  position: relative;
  top: -25px;
}
```

![Small 60px Tomster](/assets/images/posts/2022-08-05-cookbook-ember-app-to-css-modules/screen-6.png)

## Conclusion

I hope this tutorial helped you figure out the big picture about how
`ember-css-modules` isolates CSS. The way custom ids are assigned to selectors
in a given CSS file is key to understand and start working with

> `ember-css-modules`. For instance, the `composes` feature is not about

Reviewer 2: The `composes` feature has not been mentioned before, so passage is hard to understand.

Author: That's true. As the post focus on preparing the existing app for the migration, I thought maybe talking about `composes` would be out of scope. But in the other hand, that's actually quite interesting to understand what is the result of using this. So maybe it deserves a part too, or maybe I should remove that reference, even if it was tempting to mention it 🤔

Reviewer: I think you could leave it out here – it's maybe more something for a follow-up post on more advanced CSS modules aspects?

---
title: 'Cookbook: migrate an existing Ember app to CSS modules'
author: 'Marine Dunstetter'
github: BlueCutOfficial
twitter: academierenards
topic: open-source
bio: 'Senior software engineer'
description:
  'Tutorial to start with ember-css-modules in an existing Ember app. It aims at
  discovering ember-css-module from a practical perspective to get a different
  view on the doc.'
og:
  image: /assets/images/posts/2022-08-05-cookbook-ember-app-to-css-modules/og-image.png
---

This short tutorial is intended to help Ember developers to migrate an Ember app
using global CSS to CSS modules with
[`ember-css-modules`](https://github.com/salsify/ember-css-modules).

<!--break-->

## Introduction

### Prerequisites

> This article is a perfect fit for you if:

Reviewer 3: Maybe "is exactly what you are looking for" instead of "is a perfect fit"? Idk if that is used to describe articles as well, I believe it's rather meant for people :)

- You are already comfortable with Ember.

> - You know a bit of theory about what CSS modules are but you might have never

Reviewer 3: ```suggestion
- You know a bit of theory about what CSS modules are but you have never
```

  worked with them before.
- It is the first time you set up
  [`ember-css-modules`](https://github.com/salsify/ember-css-modules) in an
  existing Ember app that already contains a global CSS setup.

This article might get your interest if:

- You don't know a thing about CSS modules.
- After we've solved the previous point, you want to discover how CSS modules
  can be used in an Ember context.

### A word about CSS modules

To style an application with CSS, the minimum you need is one CSS file included
on the page. Even when split into different files, having one CSS scope that
applies globally on the whole page has some downsides in big and complex
applications. For instance, managing the styles of reusable components is
sometimes challenging. Also, the life of an application is made of turnover,
it's never easy for newcomers to identify dead code, resulting in overly heavy
and hard-to-clean CSS rules.

> CSS modules aim at answering these challenges as a technique to scope CSS

Reviewer 3: You can't really answer a challenge. Maybe "aim to overcome"?

classes locally, for instance to a single component. To read more about the
general concept, have a look at the
[documentation on Github](https://github.com/css-modules/css-modules).

### A word about `ember-css-modules`

[`ember-css-modules`](https://github.com/salsify/ember-css-modules) is a nice
addon that lets developers easily integrate CSS modules to the implementation of
their Ember apps and addon.

Installing `ember-css-modules` in an existing Ember app triggers some breaking

> changes in the way the developer should handle the app style. If the developer

Author: Do I have a missing "s" here?
```suggestion
changes in the way the developer should handle the app styles. If the developer
```

this.

### A word about `less` and `sass`

The usage of [less](https://lesscss.org/) and
[sass](https://sass-lang.com/guide) will not be mentioned in this tutorial,
though they are
[compatible with `ember-css-modules`](https://github.com/salsify/ember-css-modules/blob/master/docs/PREPROCESSORS.md).

## Migration tutorial

Let's start working. We'll take the
[SuperRentals app](https://guides.emberjs.com/release/tutorial/part-1/) as an
example. SuperRentals is the great tutorial part of the Ember guides to
introduce the main Ember features. As the purpose of the SuperRentals app is to
teach Ember and writing CSS is irrelevant to the matter, the CSS to style the
app is provided as a single `app.css` free to download (linked somewhere in the
first pages of Part 1), so newcomers don't have to write it themselves as they
walk through the tutorial.

So let's make the SuperRentals app ready for migration to CSS modules.

### Install `ember-css-modules` (and break your app)

Let's suppose we have finished implementing the SuperRentals app. The following
screenshot shows the page as it should be in your browser, with the nice styling
provided in the tutorial:

> ![SuperRental with global CSS](/assets/images/posts/2022-08-05-cookbook-ember-app-to-css-modules/screen-1.png)

Author: ```suggestion
![SuperRentals with global CSS](/assets/images/posts/2022-08-05-cookbook-ember-app-to-css-modules/screen-1.png)
```

To use CSS modules, let's install
[`ember-css-modules`](https://github.com/salsify/ember-css-modules):

```
ember install ember-css-modules
```

The addon does not contain any blueprints, the installation command simply adds
the dependency to `package.json` and modifies the lock file.

Let's run the local server again and see how our app is doing:

> ![SuperRental style is broken](/assets/images/posts/2022-08-05-cookbook-ember-app-to-css-modules/screen-2.png)

Author: ?
```suggestion
![SuperRentals styles are broken](/assets/images/posts/2022-08-05-cookbook-ember-app-to-css-modules/screen-2.png)
```

Ouch, all the styles are broken. What we see now in the browser is very similar
to a page with no CSS at all. To understand what is going on let's use the
inspector tab of the browser's debugger.

### Understanding CSS classes isolation

If we compare the previously styled app and the version we have now, we notice
that Tomster (the Ember hamster mascot) is missing at the top right of the

> screen. Let's find in the DOM the element that should display Tomster. The image

Reviewer 3: Missing the object for "let's find" :)

Reviewer 3: Or maybe you meant "let's look/search"

Author: What do you mean by the "object" for let's find? If you mean the object to find, it's "the element that should display Tomster", so maybe the whole sentence is not clear enough?

Reviewer 3: Try putting "in the DOM" at then end of the sentence then, I think it'd be clearer that way :)

In the section of the inspector dedicated to CSS description, when a style
applies to the selected element, we can see the name of the CSS file this style
comes from. On the screenshot above, it comes from `super-rental-typescript`,
which is the name of this demo application (the final "s" on "rentals" was
forgotten ^^' and it was first implemented as a migration test to TypeScript).
The file name is clickable and lets you see the content of the whole CSS file.

At the top of the file, we read:

```css
* {
  margin: 0;
  padding: 0;
}

body,
h1,
h2,
h3,
h4,
h5,
h6,
p,
div,
span,
a,
button {
  font-family: 'Lato', 'Open Sans', 'Helvetica Neue', 'Segoe UI', Helvetica,
    Arial, sans-serif;
  line-height: 1.5;
}
```

At the bottom of the file, we read:

```css
._right_1s31t7 {
  float: right;
}

._tomster_1s31t7 {
  background: url(../assets/images/teaching-tomster.png);
  background-size: contain;
  background-repeat: no-repeat;
  height: 200px;
  width: 200px;

  position: relative;
  top: -25px;
}
```

And here is the explanation about our missing Tomster. With `ember-css-modules`
installed, all the class selectors present in `app.css` end with a custom id.

> But this custom id is not present in our component's template, we still have

Reviewer 3: However, this custom id is not present in our component's template since we still have
```

Author: I am not 100% sure about the usage of "since" here. Just to clarify: there is not _really_ cause and effect link between  "this custom id is not present in our component's template" and "we still have `class="right tomster"`". It is more like two separated observations to reach the same conclusion. 
Considering this, do you think I should apply your suggestion or do you see a better way to say it?

Reviewer 3: Then "However, this custom id is not present in our component's template: we still have..."

`class="right tomster"`. As a result, no CSS class is found for `.right` and
`.tomster` and no style applies. On the contrary, the rules defined for main
elements like `div` match by the element itself and thus keep applying.

`ember-css-modules` relies on class selectors to isolate the CSS and prevent it
from applying globally. The isolation system applies to every CSS file,
including those located in the main `styles` folder rather than alongside a
component.

### Fix the setup with `:global`

> Before starting to actively use CSS modules across the app, we would prefer to

Reviewer 3: ```suggestion
Before starting to actively use CSS modules across the app, we should
```

> fix the styles first and allow `app.css` to apply globally as it did before. To

Reviewer 3: ```suggestion
fix the styles and allow `app.css` to apply globally as it did before. To
```

:global(.right) {
  float: right;
}
```

```css
/* app/components/jumbo.css */

.tomster {
  background: url(../assets/images/teaching-tomster.png);
  background-size: contain;
  background-repeat: no-repeat;
  height: 200px;
  width: 200px;

  position: relative;
  top: -25px;
}
```

```html
<!-- app/components/jumbos.hbs -->

<div class="right" local-class="tomster"></div>
```

Everything should show as before:

![Tomster still shows as expected](/assets/images/posts/2022-08-05-cookbook-ember-app-to-css-modules/screen-5.png)

If we look one more time at the CSS classes in the inspector, we'll see that
class `.right` is still applied globally, but `.tomster` selector now owns a
custom id used to isolate the CSS. The syntax
`<div class="right" local-class="tomster"></div>` we used in the template
resulted in `<div class="right _tomster_1myv6d"></div>` in the browser so the
class attributes and the generated CSS rules match.

## Conclusion

We have drawn the big picture about how `ember-css-modules` isolates CSS. The
way custom ids are assigned to selectors in a given CSS file is key to
understanding and starting to work with `ember-css-modules`. With this in mind,

> up to you to continue digging and deduce the best way to organize CSS in your

Reviewer 3: ```suggestion
it is up to you to continue digging and deduce the best way to organize CSS in your
```

---
title: 'Cookbook: migrate an existing Ember app to CSS modules'
author: 'Marine Dunstetter'
github: BlueCutOfficial
twitter: academierenards
topic: open-source
bio: 'Senior software engineer'
description:
  'Tutorial to start with ember-css-modules in an existing Ember app. It aims at
  discovering ember-css-module from a practical perspective to get a different
  view on the doc.'
og:
  image: /assets/images/posts/2022-08-05-cookbook-ember-app-to-css-modules/og-image.png
---

This short tutorial is intended to help Ember developers to migrate an Ember app

> using only one global CSS to CSS modules with

Reviewer: ```suggestion
using global CSS to CSS modules with
```

I get what you want to say but I think it doesn't quite work like that since _"only one global CSS"_ kind of implies there could also be 2 or n global CSSs :)

[`ember-css-modules`](https://github.com/salsify/ember-css-modules).

<!--break-->

## Introduction

### Prerequisites

This article is a perfect fit for you if:

- You are already comfortable with Ember.
- You know a bit of theory about what CSS modules are but you might have never
  worked with them before.
- It is the first time you set up
  [`ember-css-modules`](https://github.com/salsify/ember-css-modules) in an
  existing Ember app that already contains a global CSS setup.

This article might get your interest if:

- You don't know a thing about CSS modules.
- After we've solved the previous point, you want to discover how CSS modules

>   can be used in the Ember context.

Reviewer: ```suggestion
  can be used in an Ember context.
```

?

### A word about CSS modules

To style an application with CSS, the minimum you need is one CSS file included

> on the page. Even when split into different files, having one classic cascading
> CSS that applies globally on the whole page has some downsides in big and

Reviewer: ```suggestion
on the page. Even when split into different files, having one 
CSS scope that applies globally on the whole page has some downsides in big and
```

I think scope is the keyword here?

> complex applications. For instance, managing reusable components style is

Reviewer: ```suggestion
complex applications. For instance, managing reusable component styles is
```

Author: Just to be sure: I meant X reusable components, 1 style per reusable component.

Reviewer: Hm, maybe

```suggestion
complex applications. For instance, managing the styles of reusable components is
```

then?

sometimes challenging. Also, the life of an application is made of turnover,
it's never easy for newcomers to identify dead code, resulting in overly heavy
and hard-to-clean CSS rules.

CSS modules aim at answering these challenges as a technique to scope CSS
classes locally, for instance to a single component. To read more about the
general concept, have a look at the
[documentation on Github](https://github.com/css-modules/css-modules).

### A word `ember-css-modules`

[`ember-css-modules`](https://github.com/salsify/ember-css-modules) is a nice
addon that lets developers easily integrate CSS modules to the implementation of
their Ember apps and addon.

Installing `ember-css-modules` in an existing Ember app triggers some breaking
changes in the way the developer should handle the app style. If the developer

> is not familiar with CSS modules in the first place, they might spend a hard

Reviewer: ```suggestion
is not familiar with CSS modules in the first place, they might have a hard
```

time figuring out how to set up and start the migration at their own pace. The
present tutorial is intended to be used as a quick start cookbook to achieve
this.

### A word about less and sass

The usage of [less](https://lesscss.org/) and
[sass](https://sass-lang.com/guide) will not be mentioned in this tutorial,
though they are
[compatible with `ember-css-modules`](https://github.com/salsify/ember-css-modules/blob/master/docs/PREPROCESSORS.md).

## Migration tutorial

Let's start working. We'll take

> [SuperRentals app](https://guides.emberjs.com/release/tutorial/part-1/) as an

Reviewer: ```suggestion
the [SuperRentals app](https://guides.emberjs.com/release/tutorial/part-1/) as an
```

example. SuperRentals is the great tutorial part of the Ember guides to

> introduce the main Ember features. As the purpose of SuperRentals app is to

Reviewer: ```suggestion
introduce the main Ember features. As the purpose of the SuperRentals app is to
```

> learn Ember and writing CSS is irrelevant to the matter, the CSS to style the

Reviewer: ```suggestion
teach Ember and writing CSS is irrelevant to the matter, the CSS to style the
```

:)

app is provided as a single `app.css` free to download (linked somewhere in the
first pages of Part 1), so newcomers don't have to write it themselves as they
walk through the tutorial.

> So let's make SuperRentals app ready for migration to CSS modules.

Reviewer: ```suggestion
So let's make the SuperRentals app ready for migration to CSS modules.
```

### Install `ember-css-modules` (and break your app)

> Let's suppose we have finished implementing SuperRentals app. The following

Reviewer: ```suggestion
Let's suppose we have finished implementing the SuperRentals app. The following
```

Alternatively you could also just say "SuperRentals" without "app" but I think adding "app" requires an article.

screenshot shows the page as it should be in your browser, with the nice styling

> provided in SuperRentals tutorial:

Reviewer: ```suggestion
provided in the tutorial:
```

![SuperRental with global CSS](/assets/images/posts/2022-08-05-cookbook-ember-app-to-css-modules/screen-1.png)

To use CSS modules, let's install
[`ember-css-modules`](https://github.com/salsify/ember-css-modules):

```
ember install ember-css-modules
```

The addon does not contain any blueprints, the installation command simply adds
the dependency to `package.json` and modifies the lock file.

Let's run the local server again and see how our app is doing:

![SuperRental style is broken](/assets/images/posts/2022-08-05-cookbook-ember-app-to-css-modules/screen-2.png)

> Ouch, all the style is broken. What we see now in the browser is very similar to

Reviewer: ```suggestion
Ouch, all the styles are broken. What we see now in the browser is very similar to
```

I think using plural to refer to an app's styles is more common.

applies to the selected element, we can see the name of the CSS file this style
comes from. On the screenshot above, it comes from `super-rental-typescript`,
which is the name of this demo application (the final "s" on "rentals" was
forgotten ^^' and it was first implemented as a migration test to TypeScript).
The file name is clickable and lets you see the content of the whole CSS file.

At the top of the file, we read:

```css
* {
  margin: 0;
  padding: 0;
}

body,
h1,
h2,
h3,
h4,
h5,
h6,
p,
div,
span,
a,
button {
  font-family: 'Lato', 'Open Sans', 'Helvetica Neue', 'Segoe UI', Helvetica,
    Arial, sans-serif;
  line-height: 1.5;
}
```

At the bottom of the file, we read:

```css
._right_1s31t7 {
  float: right;
}

._tomster_1s31t7 {
  background: url(../assets/images/teaching-tomster.png);
  background-size: contain;
  background-repeat: no-repeat;
  height: 200px;
  width: 200px;

  position: relative;
  top: -25px;
}
```

And here is the explanation about our missing Tomster. With `ember-css-modules`
installed, all the class selectors present in `app.css` end with a custom id.
But this custom id is not present in our component's template, we still have

> `class="right tomster"`. As result, no CSS style is found for `.right` and

Reviewer: ```suggestion
`class="right tomster"`. As a result, no CSS class is found for `.right` and
```

`.tomster` and no style applies. On the contrary, the rules defined for main

> elements like `div` don't have any custom id added and keep applying.

Reviewer: ```suggestion
elements like `div` match by the element itself and thus keep applying.
```

I think saying there's no custom ID added to these is a bit confusing as these are element selectors and thus different from class selectors anyway. An element selector like `div-340utgrheogf` would never work anyway as that element doesn't exist.

`ember-css-modules` relies on class selectors to isolate the CSS and prevent it
from applying globally. The isolation system applies to every CSS file,
including those located in the main `styles` folder rather than alongside a
component.

### Fix the setup with `:global`

Before starting to actively use CSS modules across the app, we would prefer to

> fix the style first and allow `app.css` to apply globally as it did before. To

Reviewer: ```suggestion
fix the styles first and allow `app.css` to apply globally as it did before. To
```

do so, we can use the `:global` pseudoselector:

```css
/* app/styles/app.css */

:global(.right) {
  float: right;
}

:global(.tomster) {
  background: url(../assets/images/teaching-tomster.png);
  background-size: contain;
  background-repeat: no-repeat;
  height: 200px;
  width: 200px;

  position: relative;
  top: -25px;
}
```

And let's see what happens in the browser:

![Tomster shows again](/assets/images/posts/2022-08-05-cookbook-ember-app-to-css-modules/screen-4.png)

Tomster is back! If we take a new look at the generated CSS file in the
browser's inspector, we can see this time the classes `.right` and `.tomster`
don't have custom ids. The `:global` pseudoselector indicates these classes are

> global classes that should apply to the whole page. So we are back to a classic

Reviewer: ```suggestion
global classes that should apply to the whole page. So we are back to classic
```

CSS, with rules defined in `app.css` which apply to the page elements through

> `class` attributes. By using `:global` pseudoselector on the whole existing CSS,

Reviewer: ```suggestion
`class` attributes. By using the `:global` pseudoselector on the whole existing CSS,
```

we can style the whole app exactly as it was before installing
`ember-css-modules`, and from there we can serenely start to use some CSS
modules.

### Start using CSS modules

> In SuperRentals app, the Tomster `<div>` element is written in the `<Jumbo>`

Reviewer: ```suggestion
In the SuperRentals app, the Tomster `<div>` element is written in the `<Jumbo>`
```

So let's create the CSS file alongside the template. In the present example, the
file is `app/components/jumbo.css` as the app has a classic structure with
co-location.

It would probably make sense if the `.right` class remains a global class that
can be assigned to different elements of the page. But let's make the `.tomster`
class local:

```css
/* app/styles/app.css */

:global(.right) {
  float: right;
}
```

```css
/* app/components/jumbo.css */

.tomster {
  background: url(../assets/images/teaching-tomster.png);
  background-size: contain;
  background-repeat: no-repeat;
  height: 200px;
  width: 200px;

  position: relative;
  top: -25px;
}
```

```html
<!-- app/components/jumbos.hbs -->

<div class="right" local-class="tomster"></div>
```

Everything should show as before:

![Tomster still shows as expected](/assets/images/posts/2022-08-05-cookbook-ember-app-to-css-modules/screen-5.png)

If we look one more time at the CSS classes in the inspector, we'll see that
class `.right` is still applied globally, but `.tomster` selector now owns a
custom id used to isolate the CSS. The syntax
`<div class="right" local-class="tomster"></div>` we used in the template

> resulted to `<div class="right _tomster_1myv6d"></div>` in the browser so the

Reviewer: ```suggestion
resulted in `<div class="right _tomster_1myv6d"></div>` in the browser so the
```

class attributes and the generated CSS rules match.

> ### One last note about `class` and `local-class`

Reviewer: I'm not quite sure about this paragraph. Without saying in which case this strategy could be useful (I'm not sure which ones that are tbh), it seems you're showing an anti-pattern?

Author: I noticed it could be helpful when you start migrating a complex component, and you want to start using `local-class` in some part of the template, but another part keeps relying on global CSS as long as the component is not migrated entirely. This note is really included in the "migration in progress" idea as a way to move things at a fully controllable pace.

Author: I tried something, let me know if it sounds better.

Reviewer: I'm not sure I'm following that example tbh. In a migration like that, wouldn't you just move the entire `.rental` class over into the component and then use the local class within the component? I'm not sure why one would still want to also keep and apply the old, global class?

Author: I found it useful when playing with the testing app I used, it was just a way to split the work when you have a lot of different rules applying on elements through different selectors. But if mentioning the idea is too confusing I prefer to remove this part, because I am not sure I can find a better example, and developing this case too much would be overkill. At the begining it was more a "little tip" rather than a doc section in my head.

## Introduction

### Prerequisites

This article is a perfect fit for you if:

- You are already comfortable with Ember.
- You know a bit of theory about what CSS modules are but you might have never
  worked with them before.
- It is the first time you set up
  [`ember-css-modules`](https://github.com/salsify/ember-css-modules) in an
  existing Ember app that already contains a global CSS setup.

This article might get your interest if:

- You don't know a thing about CSS modules.
- After we've solved the previous point, you want to discover how CSS modules
  can be used in an Ember context.

### A word about CSS modules

To style an application with CSS, the minimum you need is one CSS file included
on the page. Even when split into different files, having one CSS scope that
applies globally on the whole page has some downsides in big and complex
applications. For instance, managing the styles of reusable components is
sometimes challenging. Also, the life of an application is made of turnover,
it's never easy for newcomers to identify dead code, resulting in overly heavy
and hard-to-clean CSS rules.

CSS modules aim at answering these challenges as a technique to scope CSS

> classes locally, for instance to a single component. To read more about the

Reviewer 2: ```suggestion
classes locally, for instance to individual components. To read more about the
```

general concept, have a look at the
[documentation on Github](https://github.com/css-modules/css-modules).

### A word about `ember-css-modules`

[`ember-css-modules`](https://github.com/salsify/ember-css-modules) is a nice

> addon that lets developers easily integrate CSS modules to the implementation of

Reviewer 2: ```suggestion
addon that lets developers easily integrate the CSS modules approach to the implementation of
```

This sounded as if a developer is supposed to download a few individual CSS modules and integrate them into their app.

> their Ember apps and addon.

Reviewer 2: ```suggestion
their Ember apps and addons.
```
